keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -0,0 +1,617 @@
+  EndpointSlice controller will add topology hints for each endpoint that will
+  ensure a proportional amounts are available to each zone in a cluster.
+- `Disabled`: EndpointSlice hints will not be set for this feature.
+
+When this annotation is either unspecified or not set to `Auto`, the `Disabled`
+behavior will be default. A future KEP will explore changing the default
+behavior from `Disabled` to `Auto` with a new feature gate.
+
+#### Interoperability
+
+If both this annotation and the deprecated `topologyKeys` field are set on a
+Service, `topologyKeys` will be given precedence and this annotation will be
+ignored. This will be true until the `topologyKeys` field is removed in the
+future.
+
+#### Feature Gate
+
+This functionality will be guarded by the `TopologyAwareHints` feature
+gate. This gate should not be enabled if the `ServiceTopology` gate is enabled.
+
+### API
+
+A new `EndpointHints` struct would be added to the `EndpointSlice.Endpoint`
+struct:
+
+```go
+type Endpoint struct {
+  ...
+  // hints contains information associated with how an endpoint should be
+  // consumed.
+  // +optional
+  Hints EndpointHints `json:"hints,omitempty" protobuf:"bytes,7,opt,name=hints"`
+}
+```
+
+```go
+// EndpointHints provides hints describing how an endpoint should be consumed.
+type EndpointHints struct {
+  // forZones indicates the zone(s) this endpoint should be consumed by to
+  // enable topology aware routing.
+  forZones []ForZone `json:"forZone,omitempty" protobuf:"bytes,1,name=forZones"`
+}
+```
+
+```go
+// ForZone provides information about which zones should consume this endpoint.
+type ForZone struct {
+  // name represents the name of the zone.
+  name string `json:"name" protobuf:"bytes,1,name=name"`
+}
+```
+
+
+#### Future API Expansion
+This approach would allow for future API expansion that enabled specifying
+multiple zones per endpoint with weights. That level of complexity may never be
+necessary, but it will be possible. For example:
+
+```yaml
+hints:
+  forZones:
+  - name: example-1a
+    weight: 50
+  - name: example-2a
+    weight: 50
+```
+
+Additionally we could easily expand this API to include support for region
+hints. Although it is unclear if either expansion will be necessary, the API is
+designed in a way to make expansions straightforward.
+
+### Kube-Proxy
+
+When the `TopologyAwareHints` feature gate is enabled, Kube-Proxy will be
+updated to filter endpoints based on topology hints when the following
+conditions are true:
+
+- Kube-Proxy is able to determine the zone it is running within (likely based
+  on node labels).
+- The `service.kubernetes.io/topology-aware-routing` annotation is set to
+  `Auto` for the Service.
+- At least one endpoint for the Service has a hint pointing to the zone
+  Kube-Proxy is running within.
+- All endpoints for the Service have zone hints.
+
+When the above conditions are true, kube-proxy will only route traffic to
```

Comment:
Based on the example you've added 
```
- addresses: ["10.1.2.1"]
  zone: "zone-a"
  hints:
    zone: "zone-a"
- addresses: ["10.1.2.2"]
  zone: "zone-b"
  hints:
    zone: "zone-b"
- addresses: ["10.1.2.3"]
  zone: "zone-a"
  hints:
    zone: "zone-c"
```

kube-proxies from zone A never will use 
```
- addresses: ["10.1.2.3"]
  zone: "zone-a"
  hints:
    zone: "zone-c"
```

despite it is in the same zone, is that correct?

Reply:
yeah, in this case C didn't have any so it was "donated" from A, and therefore all zones are ~equally provisioned.

Reply:
Then it is possible that pods in the same node that the endpoint, will not use it the local endpoint because was donated to other zone, I don't know how much are you into soccer/football, but there is a legendary sentence from a Brazilian coach  nicknamed Tim that says: "Playing soccer is like covering yourself with a short blanket." , that means either you cover your head, or you wrap your feet :smile:

Reply:
Haha yep, that is definitely an unfortunate downside of this approach. The previous subsetting approach was incompatible with `ExternalTrafficPolicy=Local` for the same reason.